cpp/src/io/csv/csv_writer.cu

```diff
@@ -0,0 +1,286 @@
+    unsigned int rows = (unsigned int)columns[0]->size;
+    const char* filepath = args->filepath;
+    char delimiter[2] = {',','\0'};
+    if( args->delimiter )
+        delimiter[0] = args->delimiter;
+    char terminator[3] = {'\n','\0','\0'};
+    if( args->windows_line )
+    {
+        terminator[0] = '\r';
+        terminator[1] = '\n';
+    }
+    else if( args->line_terminator )
+        terminator[0] = args->line_terminator;
+    const char* true_value = (args->true_value ? args->true_value : "true");
+    const char* false_value = (args->false_value ? args->false_value : "false");
+
+    // check for issues here
+    if( filepath==0 )
+    {
+        std::cerr << "write_csv: filepath not specified\n";
+        return GDF_INVALID_API_CALL;
+    }
+    if( count==0 )
+    {
+        std::cerr << "write_csv: num_cols is required\n";
+        return GDF_INVALID_API_CALL;
+    }
+    if( columns==0 )
+    {
+        std::cerr << "write_csv: invalid data values\n";
+        return GDF_INVALID_API_CALL;
+    }
+
+    // check all columns are the same size
+    for( int idx=0; idx < args->num_cols; ++idx )
+    {
+        gdf_column* col = args->data[idx];
+        NVStrings* strs = 0;
+        //printf("%d: %p %s %d %d\n", idx, col, col->col_name, (int)col->dtype, (int)col->size);
+        switch( col->dtype )
+        {
+            case GDF_INT8:
+            case GDF_INT32:
+            case GDF_INT64:
+            case GDF_FLOAT32:
+            case GDF_FLOAT64:
+            case GDF_DATE64:
+                if( (unsigned int)col->size != rows )
+                {
+                    std::cerr << "write_csv: columns sizes do not match (" << (int)col->size << "!=" << rows << ")\n";
+                    return GDF_COLUMN_SIZE_MISMATCH;
+                }
+                break;
+            case GDF_STRING:
+                strs = (NVStrings*)col->data;
+                if( strs->size() != rows )
+                {
+                    std::cerr << "write_csv: columns sizes do not match (" << strs->size() << "!=" << rows << ")\n";
+                    return GDF_COLUMN_SIZE_MISMATCH;
+                }
+                break;
+            default:
+                std::cerr << "write_csv: unknown column type: " << (int)col->dtype << "\n";
+                return GDF_UNSUPPORTED_DTYPE;
+        }
+    }
+
+    // check the file can be written
+    FILE* fh = fopen(filepath,"wb");
+    if( !fh )
+    {
+        std::cerr << "write_csv: file [" << filepath << "] could not be opened\n";
+        return GDF_FILE_ERROR;
+    }
+
+    //
+    // It would be good if we could chunk this.
+    // Use the rows*count calculation and a memory threshold to
+    // output a subset of rows at a time instead of the whole thing at once.
+    // The entire CSV must fit in CPU memory before writing it out.
+    //
+    // compute sizes we need
+    // build a matrix of pointers
+    int* datalens = new int[rows*count]; // matrix
```

Comment:
Please check with @vuule or @j-ieong but I believe we're modifying the CSV reader to use smart pointers so we don't have memory leaks.